### PR TITLE
[WIP][CR]Battery powered bread machine

### PR DESCRIPTION
**Pull request by Nymall**
One of the most fundamental tools in our kitchen is our bread machine. It's ridiculous how often it actually gets used in our day to day lives, as with a family of 4, bread is always a thing that's needed. Seeing as this is technology we have now, I can't imagine that in a future apocalypse all the bread machines would be gone.

**Assumptions**

- Our bread machine takes an hour and 15 minutes to run, so time has been set to 600 ticks.
- The weight of materials would not be that different than the standard bread machine.
- Most cooking operations use about 25 energy, so I assumed for that hour it would consume the same amount in batteries.
- Bread machine based off of the Hamilton Beach 29882(6.6 kg): https://www.amazon.ca/Hamilton-Beach-29882-HomeBaker-Breadmaker/dp/B005EPRF1I?source=googleshopping&locale=en-CA&tag=googcana-20&ref=pd_sl_9suozh0vg5_e
- Bread calculated as 8 "slices" at 56 grams per slice.

**Issues**

- Bread inside the machine does not rot until after it's opened. I puzzled on this or a while but couldn't figure out a way to make the bread rot but the machine not.

**ToDo**

- [x] Add autobread to loot tables.

### CODE REVIEW

File: data/json/items/tools.json [hunk 1]
```diff
@@ -8319,5 +8319,56 @@
...
+    "bashing": 8,
+    "material": [ "metal", "glass" ],
+    "symbol": "%",
+    "color": "white"
+  },
+  {
+    "id": "autobread_process",
+    "type": "GENERIC",
+    "category": "food",
+    "name": "Libotech Autobread(Churning)",
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: Space before bracket, and de-capitalize `churning`. The same for `(Done)` below.

File: data/json/items/tools.json [hunk 2]
```diff
@@ -8319,5 +8319,56 @@
...
+    "description": "This is the Libotech Autobread battery powered bread machine. It is presently in the process of making some nice, tasty bread.",
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: Two spaces after period, please.

File: data/json/items/tools.json [hunk 3]
```diff
@@ -8319,5 +8319,56 @@
...
     "color": "light_gray",
     "flags": [ "NO_SALVAGE" ]
-  }
+  },
+  {
+    "id": "autobread",
+    "type": "GENERIC",
+    "category": "food",
+    "name": "Libotech Autobread",
+    "description": "This is the Libotech Autobread battery powered bread machine. Originally marketed as a camping appliance, this machine makes coming home to a fresh loaf of bread a snap!",
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: Two spaces after period, please.

File: data/json/items/tools.json [hunk 4]
```diff
@@ -8319,5 +8319,56 @@
...
+      "not_ready_msg": "The machine is still quietly churning.",
+      "//": "1 Hour processing time"
+    }
+  },
+  {
+    "id": "autobread_done",
+    "type": "GENERIC",
+    "category": "food",
+    "name": "Libotech Autobread(Done)",
+    "description": "This is the Libotech Autobread battery powered bread machine. It has stopped churning and the smell of fresh baked bread wafts from the vents(disassemble to get bread).",
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: Two spaces after period, please. Also space before bracket. Also we use Oxford comma, so comma before `and the smell`.

File: data/json/items/tools.json [hunk 5]
```diff
@@ -8319,5 +8319,56 @@
...
     "flags": [ "NO_SALVAGE" ]
-  }
+  },
+  {
+    "id": "autobread",
+    "type": "GENERIC",
+    "category": "food",
+    "name": "Libotech Autobread",
+    "description": "This is the Libotech Autobread battery powered bread machine. Originally marketed as a camping appliance, this machine makes coming home to a fresh loaf of bread a snap!",
+    "weight": 52,
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: 52 grams for the tool?

File: data/json/items/tools.json [hunk 6]
```diff
@@ -8319,5 +8319,56 @@
...
+    "volume": 1,
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: 0,25 liters of volume for the tool?

File: data/json/items/tools.json [hunk 7]
```diff
@@ -8319,5 +8319,56 @@
...
+    "price": 10,
```
Comment (Night-Pryanik) [inline data/json/items/tools.json]: 10 cents of price for the tool?

File: data/json/recipes/recipe_food.json [hunk 8]
```diff
@@ -9722,6 +9722,23 @@
...
+    "skill_used": "cooking",
+    "difficulty": 1,
+    "time": 100,
+    "autolearn": true,
+    "components": [
+      [ [ "flour", 3 ] ],
+      [ [ "yeast", 1 ] ],
+      [ [ "water", 2 ], [ "water_clean", 2 ] ],
+      [ [ "autobread", 1] ],
+	  [ [ "battery", 25] ]
```
Comment (Night-Pryanik) [inline data/json/recipes/recipe_food.json]: Weird indentation. Also, spaces before brackets.

**Comment by Night-Pryanik** [comment]
I like the whole idea of autobread, but I don't quite like the idea of using craft mechanics to make bread with autobread as component. I understand I ask for a total rewrite, but how about adding `iuse` function to the autobread? Something similar to the `multicooker`? 
Say, `a`ctivate the autobread, check for ingredients for the bread, push the button, wait, and after some time get the bread?

**Comment by Nymall** [comment]
Ok, I'll take a look and see if I can figure out how to do it with `iuse`.

**Comment by Coolthulhu** [comment]
I'd rather prefer you went the crafting route here. Multicooker is a bunch of hacky code that is hard to generalize. I don't know your programming skills, but if you're tempted to Ctrl+c+v multicooker code, it's better that you don't. Having massive copypasted functions is a pain whenever something in them needs updating.

Extracting common function from multicooker would not be a bad idea, though.

**Comment by Nymall** [comment]
Ok, looking at the multicooker code, I might be able to extract the cooking function itself. There are three barriers I can see though:

1. The multi-cooker uses fake items to ensure the tools required appear when searching the crafting UI. This could be passed as an array, but would need a way on the item itself to track it's hidden qualities. While this could be passed in the code block, it would be better to pass it from the item itself through JSON to enable ease of modding.
2. The Multi-cooker creates two variables, but I'm not exactly sure how the game tracks and saves them. `DISH` and 'COOKTIME' are erased in the 'mc_stop' condition or if 'COOKTIME' is less than 0. I'm assuming normally this would be handled by a destructor(for saving), but I can't seem to track it down.
3. Item timers should be attached as a separate piece of code and coupled to the item in question, but I'm really not sure how to do that. What I can imagine is setting up a timeline object that pushes status changes to the item itself, and can be set up to push various messages and effects after x turns have passed. Most of my experience is with PHP, and how I'd normally handle that is create the object for the timeline and pass a pointer to the item itself(and back) and run the timeline every turn, but that seems to be not how items are set up here.

**Comment by Coolthulhu** [comment]
> The multi-cooker uses fake items to ensure the tools required appear when searching the crafting UI. This could be passed as an array, but would need a way on the item itself to track it's hidden qualities. 

There are few options:
* Have those hidden qualities hardcoded by the not-common function
* Pass them from above, as an argument
* Implement it all as an iuse actor (nicest because others could use it for their "autocrafters", but moderate amount of extra work)
* (Horrible, don't actually do it) Hardcode them by item id

> Is there a more reasonable way of handling this without creating a new Variable?

I think it's safe to reuse `item_counter`. It is context dependent (and thus kinda ugly), but bombs use it, so it is most likely safe to use for purposes not related to `HOT` and `COLD`. Assuming you only use it for one purpose at a time.

> Are there any restrictions on using vectors?

If you need to add a new field to `item` class, you are most likely doing something wrong.
You probably want to consume all the items at start of crafting and only note the recipe name in the crafter.

**Comment by BorkBorkGoesTheCode** [comment]
@Nymall ?

**Comment by Nymall** [comment]
I'm having a lot of difficulty with the code, so I think I'm going to close this pull request for now. Trying to make it not a hackfest is creating bigger and bigger issues. I'm going to revisit this in the summer when I have a little bit more free time on my hands. Sorry for the wait guys.

**Pull request closed without merging**